sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java

```diff
@@ -0,0 +1,265 @@
+    int hashcode = 1;
+    int size = 0;
+    for (int i = 0; i < sourceData.length; i += 2) {
+      int filterIndex = i / 2;
+      // If the sourceData key isn't present in includedKeys, record the exclusion in
+      // filteredIndices (or
+      // overflowFilteredIndices based on the index).
+      if (!includedKeys.contains(sourceData[i])) {
+        // Record
+        if (filterIndex < BITS_PER_INTEGER) {
+          filteredIndices = filteredIndices | (1 << filterIndex);
+        } else {
+          overflowFilteredIndices.set(filterIndex - BITS_PER_INTEGER);
+        }
+      } else { // The key-value is included in the output, record in the hashcode and size.
+        hashcode = 31 * hashcode + sourceData[i].hashCode();
+        hashcode = 31 * hashcode + sourceData[i + 1].hashCode();
+        size++;
+      }
+    }
+    // If size is 0, short circuit and return Attributes.empty()
+    if (size == 0) {
+      return Attributes.empty();
+    }
+    return new FilteredAttributes(
+        sourceData, filteredIndices, overflowFilteredIndices, hashcode, size);
+  }
+
+  private static Attributes convertToStandardImplementation(Attributes source) {
+    AttributesBuilder builder = Attributes.builder();
+    source.forEach(
+        (key, value) -> putInBuilder(builder, (AttributeKey<? super Object>) key, value));
+    return builder.build();
+  }
+
+  @Nullable
+  @Override
+  public <T> T get(AttributeKey<T> key) {
+    if (key == null) {
+      return null;
+    }
+    for (int i = 0; i < sourceData.length; i += 2) {
+      if (key.equals(sourceData[i]) && includeIndexInOutput(i)) {
+        return (T) sourceData[i + 1];
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public void forEach(BiConsumer<? super AttributeKey<?>, ? super Object> consumer) {
+    for (int i = 0; i < sourceData.length; i += 2) {
+      if (includeIndexInOutput(i)) {
+        consumer.accept((AttributeKey<?>) sourceData[i], sourceData[i + 1]);
+      }
+    }
+  }
+
+  @Override
+  public int size() {
+    return size;
+  }
+
+  @Override
+  public boolean isEmpty() {
+    // #create short circuits and returns Attributes.empty() if empty, so FilteredAttributes is
+    // never empty
+    return false;
+  }
+
+  @Override
+  public Map<AttributeKey<?>, Object> asMap() {
+    Map<AttributeKey<?>, Object> result = new LinkedHashMap<>(size);
+    for (int i = 0; i < sourceData.length; i += 2) {
+      if (includeIndexInOutput(i)) {
+        result.put((AttributeKey<?>) sourceData[i], sourceData[i + 1]);
+      }
+    }
+    return Collections.unmodifiableMap(result);
+  }
+
+  @Override
+  public AttributesBuilder toBuilder() {
+    AttributesBuilder builder = Attributes.builder();
+    for (int i = 0; i < sourceData.length; i += 2) {
+      if (includeIndexInOutput(i)) {
+        putInBuilder(builder, (AttributeKey<? super Object>) sourceData[i], sourceData[i + 1]);
+      }
+    }
+    return builder;
+  }
```

Comment:
I guess an alternative could be to implement a method that produces filtered `List<Object>` and feed the result to `ReadOnlyArrayMap.wrap` and `ArrayBackedAttributesBuilder`

Reply:
Yeah I think there are ways to optimize this but I don't believe this is on the hot path so not a priority. Can always followup.